api/v1alpha4/azurecluster_default.go

```diff
@@ -202,18 +202,18 @@ func (c *AzureCluster) setNodeOutboundLBDefaults() {
 			return
 		}
 
-		var oneSubnetWithoutNatGateway bool
+		var needsOutboundLB bool
 		for _, subnet := range c.Spec.NetworkSpec.Subnets {
 			if subnet.Role == SubnetNode && !subnet.IsNatGatewayEnabled() {
-				oneSubnetWithoutNatGateway = true
+				needsOutboundLB = true
 				break
 			}
 		}
 
 		// If we don't default the outbound LB when there are some subnets with nat gateway,
 		// and some without, those without wouldn't have outbound traffic. So taking the
 		// safer route, we configure the outbound LB in that scenario.
-		if len(c.Spec.NetworkSpec.Subnets) > 0 && !oneSubnetWithoutNatGateway {
+		if !needsOutboundLB {
```

Comment:
@CecileRobertMichon `TestNodeOutboundLBDefaults` is failing because it is implemented to expect a load balancer object even if there are no subnets in the input spec. (See the removal of the length requirement above.)

Is that a valid way to express a UT? Will a cluster ever not actually have any subnets by the time that `setNodeOutboundLBDefaults` is entered into?

We could either restore that length requirement (doesn't make sense to me or I could be missing something) or fix up the existing UT input/output for the default case to express what we expect given this change.

Reply:
> Will a cluster ever not actually have any subnets by the time that setNodeOutboundLBDefaults is entered into

short answer is no, we set subnet defaults before setting LB defaults on line 57. The length requirement doesn't make sense to me either, if there are no subnets (that wouldn't happen but in theory) then I don't see why we need an outbound LB added. I vote we change the UT to make sure meaningful edge cases are covered instead: all subnets have a NAT GW, none have it, all but one, and all node subnets have it but not the control plane subnet.

Reply:
in terms of the "all subnets have a NAT GW", is that real scenario? will we ever have a cluster that doesn't have a control plane subnet, and/or do we support a control plane subnet that uses NAT GW?

Reply:
you're right, my bad. NAT GW is only for node subnets. So all, some, none of the node subnets have NAT GW needs to be tested.

Reply:
O.K. cool, thanks!

Reply:
One more question on the "any one node subnet does not have NAT Gateway enabled" scenario. Currently, we seem to configure a single "default flavor" of outbound LB. Does that mean if any one node pool (assuming a 1:1 mapping of node pools to subnets) has NAT Gateway disabled, that we will create a single outbound LB that has *every single node in the cluster* in its backend pool? The logic as is doesn't seem smart enough to only use the outbound LB for node VMs that aren't configured for NAT Gateway.

Hope that makes sense. Basically, as I understand it, the current "some but not all nodes are using NAT Gateway" implementation seems sort of half-baked (if it even works at all).

Reply:
See  https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1411#discussion_r670578995

Reply:
cc @fiunchinho

Reply:
Thanks.

In the scenario where we have _n_ node pools with NAT Gateway and _n_ without, then essentially we create the outbound LB so that all nodes have outbound SNAT (not just those pools with NAT Gateway enabled). But in that case we don't actually use the NAT Gateway in practice, because the outbound LB has precedence over the NAT Gateway for all backend pool members that have a NAT Gateway (in our configuration the overlap is 100% — *all* of the NAT Gateway-enabled nodes will be members of this outbound LB).

If the above is correct, then I'm not sure why we support things in this way. It would make more sense to require the user "choose one or the other", because otherwise the user is effectively paying for a NAT Gateway that is not being used. (among other undesirable outcomes)

Reply:
If I understand correctly, that's the end goal with https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1532 . After that is fixed, we would only support Nat Gateways for outbound traffic from the node subnets.

Reply:
Does that include control plane nodes? So the idea is we don't re-use the control plane LB for outbound SNAT (assuming we're in a non-private cluster scenario where there _is_ a LB for control plane nodes) and instead use a dedicated NAT gateway for the control plane nodes?

Reply:
For control plane nodes there are two cases:

- Public API Server (not a private cluster): the API Server LB also provides outbound, no need for another outbound mechanism
- Private cluster: The internal API Server LB does not provide outbound, we currently deploy a separate control plane outbound LB just for outbound. We could also switch this to instead creating a NAT gateway for the control plane subnet, but the SNAT port exhaustion issue isn't as much of a problem here since scaling control plane nodes to 60 VMs isn't a requirement.